File: scripts/classes/dmi.lua
--- Dmi is a table representing a DMI file.
--- @class Dmi
--- @field name string The name of the DMI file.
--- @field width number The width of the DMI file.
--- @field height number The height of the DMI file.
--- @field states (State)[] The states of the DMI file.
--- @field temp string The temporary directory of the DMI file.
Dmi = {}
Dmi.__index = Dmi

--- Creates a new Dmi object.
--- @param dmi table The Dmi table containing the properties of the Dmi object.
--- @return Dmi dmi The newly created Dmi object.
function Dmi.new(dmi)
	local self = setmetatable({}, Dmi)

	self.name = dmi.name
	self.width = dmi.width
	self.height = dmi.height
	self.states = {}
	self.temp = dmi.temp

	for _, state in ipairs(dmi.states) do
		self.states[#self.states + 1] = State.new(state)
	end

	return self
end

--- Represents a state object.
--- @class State
--- @field name string The name of the state.
--- @field dirs 1|4|8 The number of directions in the state.
--- @field frame_key string The frame key of the state.
--- @field frame_count number The number of frames in the state.
--- @field delays (number)[] The delays of the state.
--- @field loop_ number Whether the state loops or not.
--- @field rewind boolean Whether the state rewinds or not.
--- @field movement boolean Whether the state is a movement state or not.
--- @field hotspots (string)[] The hotspots of the state.
State = {}
State.__index = State

--- Creates a new State object.
--- @param state table The state table containing the properties of the State object.
--- @return State state The newly created State object.
function State.new(state)
	local self = setmetatable({}, State)

	self.name = state.name or ""
	self.dirs = state.dirs or 1
	self.frame_key = state.frame_key or ""
	self.frame_count = state.frame_count or 1
	self.delays = {}
	self.loop_ = state.loop_ or 0
	self.rewind = state.rewind or false
	self.movement = state.movement or false
	self.hotspots = {}

	for _, delay in ipairs(state.delays) do
		self.delays[#self.delays + 1] = delay
	end

	for _, hotspot in ipairs(state.hotspots) do
		self.hotspots[#self.hotspots + 1] = hotspot
	end

	return self
end

File: lib/tests/dmi.rs
use std::fs::{remove_dir_all, remove_file};
use std::path::Path;

use dmi::Dmi;

#[test]
fn open_and_save() {
    let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
    let sample_file = manifest_dir.join("tests/assets/anomaly.dmi");

    let dmi = Dmi::open(sample_file).unwrap();

    assert_eq!(dmi.name, "anomaly");
    assert_eq!(dmi.width, 32);
    assert_eq!(dmi.height, 32);
    assert_eq!(dmi.states.len(), 9);

    let state = &dmi.states[0];

    assert_eq!(state.name, "carp_rift");
    assert_eq!(state.dirs, 1);
    assert_eq!(state.frames.len(), 3);
    assert_eq!(state.frame_count, 3);
    assert_eq!(state.delays.len(), 3);
    assert_eq!(state.loop_, 0);

    assert!(!state.rewind);
    assert!(!state.movement);

    assert_eq!(state.hotspots.len(), 0);

    let frame = &state.frames[0];

    assert_eq!(frame.width(), 32);
    assert_eq!(frame.height(), 32);

    let delay = &state.delays[0];

    assert_eq!(delay, &2.0);

    let temp_dir = Path::new(env!("CARGO_TARGET_TMPDIR"));
    let temp_file = temp_dir.join("anomaly.dmi");

    if let Err(e) = dmi.save(&temp_file) {
        let _ = remove_file(temp_file);
        panic!("{e:?}");
    }

    let _ = remove_file(temp_file);
}

#[test]
fn serialize_and_deserialize() {
    let dmi = Dmi::open("tests/assets/anomaly.dmi").unwrap();

    let temp_dir = Path::new(env!("CARGO_TARGET_TMPDIR"));

    let serialized = dmi.to_serialized(temp_dir, false).unwrap();

    let temp = serialized.temp.clone();

    let deserialized = Dmi::from_serialized(serialized);

    let Ok(deserialized) = deserialized else {
        let _ = remove_dir_all(temp);
        panic!("{:?}", deserialized.unwrap_err());
    };

    let _ = remove_dir_all(temp);

    assert_eq!(dmi.name, deserialized.name);
    assert_eq!(dmi.width, deserialized.width);
    assert_eq!(dmi.height, deserialized.height);
    assert_eq!(dmi.states.len(), deserialized.states.len());

    let state = &dmi.states[0];
    let deserialized_state = &deserialized.states[0];

    assert_eq!(state.name, deserialized_state.name);
    assert_eq!(state.dirs, deserialized_state.dirs);
    assert_eq!(state.frames.len(), deserialized_state.frames.len());
    assert_eq!(state.frame_count, deserialized_state.frame_count);
    assert_eq!(state.delays.len(), deserialized_state.delays.len());
    assert_eq!(state.loop_, deserialized_state.loop_);
    assert_eq!(state.rewind, deserialized_state.rewind);
    assert_eq!(state.movement, deserialized_state.movement);
    assert_eq!(state.hotspots.len(), deserialized_state.hotspots.len());

    let frame = &state.frames[0];
    let deserialized_frame = &deserialized_state.frames[0];

    assert_eq!(frame.width(), deserialized_frame.width());
    assert_eq!(frame.height(), deserialized_frame.height());

    let delay = &state.delays[0];
    let deserialized_delay = &deserialized_state.delays[0];

    assert_eq!(delay, deserialized_delay);
}
